beetsplug/metasync/__init__.py

```diff
@@ -14,28 +14,72 @@
 
 """Synchronize information from music player libraries
 """
+from abc import abstractmethod, ABCMeta
+import inspect
+import pkgutil
+from importlib import import_module
 
-from beets import ui, logging
+from beets.util.confit import ConfigValueError
+from beets import ui
 from beets.plugins import BeetsPlugin
-from beets.dbcore import types
-from beets.library import DateType
-from sys import modules
-import inspect
 
-# Loggers.
-log = logging.getLogger('beets.metasync')
+
+METASYNC_MODULE = 'beetsplug.metasync'
+
+
+class MetaSource(object):
+    __metaclass__ = ABCMeta
+
+    def __init__(self, config, log):
+        self.item_types = {}
+        self.config = config
+        self._log = log
+
+    @abstractmethod
+    def sync_from_source(self, item):
+        pass
+
+
+def load_meta_sources():
+    """ Returns a dictionary of all the MetaSources
+    E.g., {'itunes': Itunes} with isinstance(Itunes, MetaSource) true
+    """
+
+    def is_meta_source_implementation(c):
+        return inspect.isclass(c) and \
+            not inspect.isabstract(c) and \
+            issubclass(c, MetaSource)
+
+    meta_sources = {}
+
+    module_names = [name for _, name, _ in pkgutil.walk_packages(
+        import_module(METASYNC_MODULE).__path__)]
+
+    for module_name in module_names:
+        module = import_module(METASYNC_MODULE + '.' + module_name)
+        classes = inspect.getmembers(module, is_meta_source_implementation)
+
+        for cls_name, _cls in classes:
+            meta_sources[cls_name.lower()] = _cls
+
+    return meta_sources
```

Comment:
This is some very clever work to get auto-discovery in order! It could be, though, that it would be easier just to hard-code the list of modules to import and pull classes out of—this sort of `pkgutil` and `inspect` magic could make this harder to maintain in the future. (We already run into some amount of trouble with testing, for example, when it interacts with similar magic in the beets plugin loading system.)

So I'll advocate gently for the "dumb" solution here: just redundantly writing down the list of available backends rather than automatically discovering them. But if anyone thinks this is too nifty to throw away, I could be convinced. :smiley: